### DISCUSSION TITLE
possible future features

### DISCUSSION THREAD

**Discussion by Kurdakov** [ThioJoe/YT-Spammer-Purge#279] [Ideas]
I think current utility is more or less feature complete (mostly I use hold action on spammer ids), if only youtube made some more efforts to stop spamming....

First feature I could think about is  to compare channel user profile picture and spammer profile picture (quite often scammers use channel owners profile picture) - it's possible to request profile pictures via api and there are some libraries to compare images (the fast one is https://github.com/JohannesBuchner/imagehash
from PIL import Image
import requests
from io import BytesIO
import imagehash
...
  get thumbnail url - channels list snippet.thumbnails.default.url
...
response = requests.get(url)
ownerImg = Image.open(BytesIO(response.content)) 
or
ownerImg = Image.open(requests.get(url, stream=True).raw)
ownerHash = imagehash.average_hash(ownerImg)
...
tesHash = imagehash.average_hash(testUserImg)

if (ownerHash == testHash):
    #same profile image found on user account
 ), one complication here - rarely legit users assume channel owner user profile picture (but then, they should know possible consequences, also from channels it's possible to get channel creation date snippet.publishedAt - recently created profiles are more probable scammers)

If someone creates web app interface with either Flask or Django  what comes to mind is to use service credentials
https://developers.google.com/identity/protocols/oauth2/service-account the problem is youtube won't support 
https://developers.google.com/youtube/v3/guides/authentication a link of service credentials to user account

(so it would not be possible to ask users of utility to just add some account as their moderators) 
but seems there is a way around (an old C# question on stackoverflow https://stackoverflow.com/questions/64975887/google-oauth2-authorization-on-linux-headless-server-net, possibly this piece of code should be tried first before moving to develop headless webapp )

if no additional quota is granted, possibly scanning should be implemented via http requests and only hold actions are performed via api ( deletion of 10 000 spammer posts per day is not bad  after all )

there are all sorts of complications with the approach (youtube denying access for server, spammers attacking server with DoS attacks, users not adding account to moderators, etc ) but still potentially it might make the utility to be much more widely used by channel owners (the use of service might also include provision to make reference to author channel or utility or making donation for server support and maintenance etc or pay for monitoring channel for existing spammers (or spammers with channel owners profile pictures).

**Comment by Kurdakov** [comment]
another idea - if someone wrote tutorial how to use YoutubeSpammerPurge on Android with Termux will possibly extend user base.
something along the way:

install Termux
in termux
termux-setup-storage
exit
in termux
pkg update
pkg upgrade

pkg install python

extract code to some directory say top level directory YouTubeSpammerPurge along with Download folder (using for ex Total Commander or Solid Explorer ):
or if one does not want install filebrowser
pkg install curl
cd /storage/emulated/0/
TAG =  curl https://api.github.com/repos/ThioJoe/YT-Spammer-Purge/releases/latest -s | jq .name -r
and then
curl https://codeload.github.com/ThioJoe/YouTube-Spammer-Purge/tar.gz/refs/tags/v$TAG -o yt-spammer.tar.gz
tar -xzf yt-spammer.tar.gz
rm yt-spammer.tar.gz
cd /storage/emulated/0/YouTube-Spammer-Purge-$TAG/'
or if extracted to some directory say YouTubeSpammerPurge with filebrowser
cd /storage/emulated/0/YouTubeSpammerPurge

pip install -r requirements.txt

pkg install python-tkinter

python YouTubeSpammerPurge.py

(btw for Android users much shorter name would be preferable (but there is import in gui.py, so it is not just about renaming YouTubeSpammerPurge.py, change in gui.py is required )

Termux should be always exited (or it will continue to run)

ps maybe a warning should be that someone needs to feel very confident Android user before making all these steps

**Reply by svgPhoenix** [reply]
> another idea - if someone wrote tutorial how to use YoutubeSpammerPurge on Android with Termux will possibly extend user base. something along the way:
> 
> install Termux in termux termux-setup-storage exit in termux pkg update pkg upgrade
> 
> pkg install python
> 
> extract code to some directory say top level directory YouTubeSpammerPurge along with Download folder (using for ex Total Commander or Solid Explorer ): or if one does not want install filebrowser pkg install curl cd /storage/emulated/0/ TAG = curl https://api.github.com/repos/ThioJoe/YT-Spammer-Purge/releases/latest -s | jq .name -r and then curl https://codeload.github.com/ThioJoe/YouTube-Spammer-Purge/tar.gz/refs/tags/v$TAG -o yt-spammer.tar.gz tar -xzf yt-spammer.tar.gz rm yt-spammer.tar.gz cd /storage/emulated/0/YouTube-Spammer-Purge-$TAG/' or if extracted to some directory say YouTubeSpammerPurge with filebrowser cd /storage/emulated/0/YouTubeSpammerPurge
> 
> pip install -r requirements.txt
> 
> pkg install python-tkinter
> 
> python YouTubeSpammerPurge.py
> 
> (btw for Android users much shorter name would be preferable (but there is import in gui.py, so it is not just about renaming YouTubeSpammerPurge.py, change in gui.py is required )
> 
> Termux should be always exited (or it will continue to run)
> 
> ps maybe a warning should be that someone needs to feel very confident Android user before making all these steps

The problem with using Termux is that, once you get as far as you've written and actually run the program with a client_secrets file present, it can't launch a web browser to complete OAuth

**Reply by Kurdakov** [reply]
somehow it launched OAuth at my phone when I tested

**Reply by svgPhoenix** [reply]
what browser are you using?

**Reply by Kurdakov** [reply]
Google chrome for Android,
maybe https://github.com/ThioJoe/YT-Spammer-Purge/wiki/Instructions:-Obtaining-an-API-Key point 14 is needed?

**Reply by svgPhoenix** [reply]
No, that error would appear in the browser. I use Firefox on a Samsung so Google Chrome isn't installed. I'll try it out in a bit.

**Reply by Kurdakov** [reply]
I remember I also enabled some permissions, but cannot recall which. Also maybe enabling developer mode on phone might change something

**Comment by Firecul** [comment]
~~The biggest problem with your image compare solution (I'm guessing from the name) is that that would only work if the image hashes are the same no? One channel I follow has a lot of impersonators but they don't use *identical* pictures. They are slightly different jpg artefacts and in some cases old channel pictures.~~

**Reply by Kurdakov** [reply]
> The biggest problem with your image compare solution (I'm guessing from the name) is that that would only work if the image hashes are the same no? One channel I follow has a lot of impersonators but they don't use _identical_ pictures. They are slightly different jpg artefacts and in some cases old channel pictures.

no, the ImageHash is specifically made to account image difference, so it's possible to set distance (like in Levenshtein comparing words) and it will account for artifacts and alterations (using older channel images is a problem, though)

**Comment by nick12357** [comment]
> pictures.

We could use pixels instead so if the pixels are a bit different it would still block the pfp.
